README
OVA (Ogg Vorbis Archive)  
Originally hosted on http://ova.sourceforge.net  
Author: Gene Pavlovsky <gene.pavlovsky@gmail.com>

This is an old project of mine that is still usable but is dated.
One day I might decide to put some more love in it.

1. Introduction

Helps to maintain an archive of music in the Ogg Vorbis format.

Main features:

  * CD ripping (front-end to cdparanoia)
  * MP3/wave -> Ogg Vorbis encoding (front-end to oggenc and mplayer for MP3s)
  * CDDB querying
        o given a CD (or saved CD discid)
        o given a bunch of MP3/Ogg Vorbis/wave files (very useful for downloaded albums)
  * identify albums/tracks interface
        o getting tracks' information (artist, album, titles etc.) from CDDB or
              + pathnames (using elaborate and customizable sed patterns)
              + tags (for MP3/Ogg Vorbis)
        o manual editing of tracks' information
              + interactive
              + in your favorite editor
        o saving the final information
  * tag&relocate tracks interface
        o tagging Ogg Vorbis files with the saved information
        o renaming and moving them according to elaborate and customizable shell function
        o associated files/subdirectories (lyrics, covers etc.) are moved automatically
  * vorbisgain (ReplayGain for Ogg Vorbis) support
  * splitting single-file albums to separate tracks
  * CDDB DTITLE database (requires local CDDB archive)
        o slow REGEXP search
        o creating of artist and album search indices
        o fast leading substring search for artist/album using search indices

2. Documentation

2.1. Installation

For the installation instructions and the list of required
software, refer to the INSTALL file.

2.2. Shell scripts

Config files: 'global.conf' (used by all shell scripts)

2.2.1. ova

This is the main script, providing dialog-based interface to the most of
the functionality.

Apart from the usual '--help' and '--version', ova doesn't accept options.
When run, ova will display the main menu with the list of available actions.
When some action is completed or aborted, ova will return to the main menu.

Actions:

2.2.1.0. 0: split single-file albums

In short:

Some people rip CDs to single MP3/wave files. Knowing the album's
CDDB category and discid, it's possible to split such files back to
separate tracks.

Notes:

By default, all tracks longer than 'track_split_playlength' value
in 'music.conf' are selected.

2.2.1.1. 1: rip an audio compact disc

Notes:

'tracks to rip' prompt: enter the rip range

A couple of examples should clarify the rip range format:

tracks to rip: -
  rip all tracks (missing endpoints mean the first and the last tracks)
tracks to rip: -3 5-7 9 14-
  rip tracks 1,2,3,5,7,9, and from track 14 to the last track

Empty rip range tells ova to quit to the main menu.

Features:

- Doesn't rip tracks that are already fully ripped (or encoded).
- Size, average speed and rip time display for previous track.
- The cdparanoia progress display for current track (as-is).

Misfeatures:

- Elapsed time is only updated after each track, and I'm not
motivated to improve it.

2.2.1.2. 2: re/encode tracks to ogg

In short:

Encode all wave files found in 'selected directory' to Ogg Vorbis.
Decode all MP3 files found in 'selected directory' to wave files,
and encode them to Ogg Vorbis. Remove all source files on success.

Notes:

The number of simultaneous jobs can be changed in 'main.conf'.
Various oggenc options (e.g., quality) can be changed in the
'oggenc_set_flags' function in 'conv.conf'.

Features:

- Total progress, elapsed and estimated time display.
Unfortunately, estimated time drifts quite a while when encoding
a large number of tracks. This is due to the low precision with
which total progress is calculated (due to shell limitations).
- Per-job filename, status, progress, elapsed and estimated time display.
- SIGQUIT (ctrl + '\') empties the queue. As soon as already running jobs
will finish, encoding will stop.

2.2.1.3: 3: identify whole albums

In short:

For each directory != 'misc' in 'selected directory', do the following:
If the 'discid'/'discid~' file exists, query the CDDB using it, otherwise
query the CDDB using files' play lengths. Allow the user to choose from
the CDDB matches, or to manually enter the CDDB category and discid.
If no CDDB entry is chosen, get information from tags and pathnames.
Allow to further edit the information manually (interactively or in
the editor). Save the final information to '.tag' files. The 'discid~'
file is removed if present, the 'discid' is kept if present. If
'cddb_submit_url' in 'cddb.conf' is set, save the CDDB entry for
later submission.

For all 'misc' directories ('misc' or 'genre/misc') in 'selected
directory', get information from tags and pathnames. Allow to
further edit the information manually. Save the final information
to '.tag' files. The 'misc' directories keep completely sundry
tracks. If you want to classify them by genre (per-genre 'misc'
directories), do it by hand first, moving all same-genre tracks
to separate directories (preferrably genre name). Afterwards,
in ova:identify, when editing the information manually, you
may set the genre if it was not auto-detected from directory.

Notes:

CDDB matches selection:
  - 'm': Enter the CDDB category and discid manually.
  - 'v': For the rare cases of 'various artists' misdetection, toggle
    it to force 'yes' or 'no' (or return to the default 'auto').

Tags and pathnames guesswork (appears if the CDDB step was unsuccessful):
  - '3': Attempt to match pathnames with patterns from 'pathname_patterns'
    array set in 'music.conf'. Patterns are tried best-to-worst. At any
    rate, the pattern can be specified manually.

Manual adjustments:
  - '3': Edit the current info in your editor of choice. The editor is
    selected by setting the 'EDITOR' environment variable. Current info
    is dumped into a temporary file in the form of a shell script, and
    is executed as a shell script after editing; if any interpretation
    errors occur, the error message is displayed, but some part of this
    erroneous script is probably already evaluated. If in doubt, better
    restart identifying this directory from scratch.
  - '4': Swap trackartist with trackalbum for each track. Useful for
    CDDB entries in which some dimwit mixed them up.
  - '5': Convert all info from the selected character set. Useful to
    convert info obtained from tags that are not in your locale's
    encoding. Also may be useful for info read from CDDB: freedb now
    endorses UTF-8, but in case you encounter entries in other encodings,
    choose the right encoding (i.e., cp1251) and the information will
    be converted from it.

The 'album' and 'disc#' prompt: The album name is automatically scanned
for something like '(cd1)' or '[disc 2]', but it can't be 100% sure,
so it's necessary to allow the user to enter (or remove misdetected)
'disc#' manually (and edit the 'album' accordingly).

2.2.1.4. 4: tag&relocate tracks

In short:

For each directory != 'misc' in 'selected directory', do the following:
Tag all Ogg Vorbis files for which the '.tag' files were found.
Relocate these Ogg Vorbis files according to the name produced by the
'track_path_pattern' function in 'music.conf'. Remove '.tag' files.
If all Ogg Vorbis files were moved to the same directory, move all
files and directories without Ogg Vorbis files from the source directory
to the target directory. Also, if the parent of the source directory
(and it's children) doesn't contain any Ogg Vorbis files, move
everything from there to the parent of the target directory. Remove
the empty directories left (if any).

For all 'misc' directories ('misc' or 'genre/misc') in 'selected
directory', tag all Ogg Vorbis files for which the '.tag' files were
found. Relocate these Ogg Vorbis files according to the name produced
by the 'track_path_pattern' function in 'music.conf'. Remove '.tag'
files. Remove the empty directories left (if any).

2.2.1.5. h: update hard links

In short:

Front-end to "ova-hardlink -a", with progress, elapsed/estimated/total
times and last processed files/directories display. For information
on ova-hardlink, see 2.2.11. Don't miss it, it's cool stuff (IMAO).

2.2.1.6. v: calculate replay gain

In short:

For each directory != 'misc' in 'music root' except the 'incoming
directory', do the following:
For all Ogg Vorbis files without REPLAYGAIN comments, calculate
and save track and album replay gain values (vorbisgain -f -a).

For all 'misc' directories in 'music root' except the 'incoming
directory', do the following:
For all Ogg Vorbis files without REPLAYGAIN comments, calculate
and save track replay gain values (vorbisgain -f).

2.2.1.7. s: CDDB submit

In short:

Submit the pending CDDB entries.

Notes:

For the CDDB server to accept your submissions, you should set
'cddb_submit_user_email' in 'cddb.conf' to a valid e-mail address.

The CDDB submit is performed via HTTP, using wget.
If you want to use HTTP proxy, configure wget to do so
(e.g., set http_proxy in wgetrc config file).

2.2.1.8. cs: cleanup stale files

In short:

Remove error files and unfinished Ogg Vorbis/wave files left
after unclean 'encode' termination.
Remove all error log files but the current.

2.2.1.9. cc: cleanup CDDB cache

In short:

Remove locally cached CDDB entries.

2.2.1.10. cd: cleanup discid list

In short:

Remove the list of ripped CDs' discids.

Notes:

This list is used for checking if the CD had already been ripped.

Only do it if your discid list becomes stale, because it's not very fast
to generate a new list (which will happen upon the next CD rip).

2.2.1.11. i: information

In short:

Display the size of the local CDDB cache, and the number of
the pending CDDB submissions.

2.2.1.12. Compound actions

There're several compound actions available.
  - 12: run 1, 2 in sequence
  - 1234: run 1, 2, 3, 4 in sequence
  - 234: run 2, 3, 4 in sequence
  - 34: run 3, 4 in sequence

Also, all compound actions ending with '4' may be appended with 'v'
which means ReplayGain (vorbisgain) calculation. The '4v' compound
action is also acceptable.

2.2.1.13. Notes

NB: Before re/encoding your entire music collection, be sure that you
have splitted all the single-MP3/wave albums to separate tracks.

Signals: the INT, TERM, HUP and QUIT signals are blocked most of the
time. During lengthy actions like encoding or ripping, the INT signal
is handled as rather-clean interrupt.

Config files: 'music.conf', 'cddb.conf', 'main.conf'

2.2.2. ova-cddb_ibuild, ova-cddb_isearch

These scripts allow to build and search the CDDB DTITLE database and
search indices. To build this database you need a complete copy of CDDB.
If you have it installed under '/usr/media/freedb', the following command
will build the database itself and the search indices:

ova-cddb_ibuild --save-db /usr/media/freedb

You can also do it this way:

ova-cddb_ibuild --save-db --no-indices /usr/media/freedb
ova-cddb_ibuild --load-db

For Nov 2003 freedb database, it took ~220 minutes on my Athlon-1400,
with ReiserFS filesystem (mounted loopback) on Seagate 7200.7 HDD.

I've implemented the 'resume' feature in ova-cddb_ibuild allowing
to interrupt the indices creation process at any time and to continue
it later, but, due to apparent bug(s) in handling signals in bash,
I had to disable this feature. The bug manifests itself as the code
assigned to handle signals doesn't get executed from time to time.
The same problem sometimes occurs when interrupting ova actions.
Also, sometimes currently evaluating shell/subshell expressions or
running child processes (e.g., mplayer/oggenc when encoding) receive
the signal, potentially leading to very bad consequences.

The ova-cddb_isearch script allows to grep through the DTITLE
database for any REGEXP (it's slow, though) using '--dtitle' option,
or to search for the leading substring of artist/album using
'--artist'/'--album' options, with optional secondary matching using
'--with-album'/'--with-artist' options.

Config files: 'cddb.conf'

2.2.3. ova-cddb_match

This script reads a list of CDDB entries from the standard input,
optionally matches them by several criteria, and display the disc
title, disc length, tracknumber, and, if given '--print-info' option,
the track title list. Matching criteria can be specified as options
('--length', '--tracks', '--ttitle'), or given a directory or file.
If a directory is given, length is matched to all it's tracks length,
track count is matched to it's track count. If a file is given,
length is matched to it's length, track count is matched to 1. To
disable matching some of the criteria, use '--length'/'--tracks'
options with empty argument. Very useful in conjunction with
ova-cddb_isearch (e.g., when you know the artist but don't
know the album, and the tracks are not numbered).

Example:

ova-cddb_isearch --artist=nightwish --with-album=oceanborn |
  ova-cddb_match /usr/media/music/incoming/rock/nightwish/oceanborn

The tracks in this album only had titles, so it wasn't possible to
do a CDDB query using their play lengths. The ova-cddb_isearch
with the aforementioned options returns 23 results (many albums
are released in different versions for various countries, with
different number and/or ordering of tracks sometimes; there are
live versions etc.) which is too much. After ova-cddb_match
there are only five CDDB entries left (sorted out by disc length
and tracks count) which is quite ok.

Config files: 'cddb.conf'

2.2.4. ova-cddb_number

This script takes further the example from the previous script.
Being passed the CDDB entry (category and discid) on the command line,
as well as the directory, it attempts to match (quite fuzzy) track
titles read from the CDDB with the names of files in this directory,
and for successful matches, renames the files by prepending track
numbers to them.

Example:

ova-cddb_number rock 8a0c740b /usr/media/music/incoming/rock/nightwish/oceanborn

The '--dry-run'/'-n' option allows to preview what would have been
done before actually doing it. The CDDB category and discid were
taken from the best match from the previous example (best by the
terms of disc length difference).

Config files: 'cddb.conf'

2.2.5. ova-cddb_query, ova-cddb_read

These two scripts do a CDDB query and a CDDB read, respectively.
The ova-cddb_query script allows quite a wide variety of input
to construct it's query. It can be given a file named 'discid' or
'discid~' which should contain the CD discid in cddb protocol
'cddb query' command format. It can be given a list of files from
play lengths of which to compute the discid. Finally, it can be
given a list of track frame offsets from which to compute the discid.
The ova-cddb_read just reads the given CDDB entry.
Both scripts can print the full discid, the list of track frame
offsets, the disc length and the track count ('-d', '-o', '-l'
and '-t' options, respectively).

Examples:

ova-cddb_query /usr/media/music/incoming/cdda/0001/discid
ova-cddb_query unknown_album1/*mp3
find unknown_album2 -maxdepth 1 -type f | sort | ova-cddb_query --filelist=-
ova-cddb_read newage a3112c0d

Notes:

The CDDB query/read is performed via HTTP, using wget.
If you want to use HTTP proxy, configure wget to do so
(e.g., set http_proxy in wgetrc config file).

Config files: 'cddb.conf'

2.2.6. ova-conv

This script converts MP3/wave file to Ogg Vorbis file, using
mplayer and oggenc. It also validates play lengths after each
conversion step (they shouldn't change for more than the value
of 'encdec_deviation' in 'global.conf').
Not very useful by itself, it's the primary worker inside
ova in encode action.

Config files: 'conv.conf'

2.2.7. ova-gettag

This script creates a '.tag~' file for a given file.
Not very useful by itself, it's heavily used inside ova
here and there.

Config files: 'music.conf'

2.2.8. ova-lint

This script searches for missing tracks in the middle of albums
(e.g., if album has tracks 1,2,3 and 6,7,9, it'll report that
tracks 4-5 and 8 are missing). Unless '--no-tags' option is
specified, it'll also check for files that miss some tag info.

Notes:

It works best on already identified, tagged&relocated tracks.
As it's designed exactly for this, you probably shouldn't use it
on non-ova'd tracks (e.g., ones in $music_incoming).

Config files: 'music.conf'

2.2.9. ova-cue2offsets

This script reads a given cue sheet, and outputs track frame offsets
file suitable for input into tracksplit.

2.2.10. ova-sed

Reads a given .ogg files' comments (tags), processes them with sed
(sed's "-e SCRIPT", "-f FILE" and "-r" options are available), and
then saves the resulting comments back to the file. New comments
can be added with "-a COMMENT" option. The script also tries to
apply sed script(s) to filename, if it's changed, move the old file
to the new filename. Combined with a little shell scripting, ova-sed
allows to quickly change, for example, the genre of multiple .ogg
files.

2.2.11. ova-hardlink

Makes hard links for a given album (processes all albums with "-a" option),
according to hardlink.conf. This allows to have several different paths to
music files while using little additional disk space.

The default setup creates these hard links:
  by album
    /usr/media/music-by/album/ak_note/akiko_kohara/2003-ak_note
  by artist
    /usr/media/music-by/artist/akiko_kohara/2003-ak_note
  by date
    /usr/media/music-by/2003/akiko_kohara/2003-ak_note
  by date and genre
    /usr/media/music-by/2003/acid_jazz/akiko_kohara/2003-ak_note
  by genre and date
    /usr/media/music-by/acid_jazz/2003/akiko_kohara/2003-ak_note
  by the most recently modified file's modification time
    /usr/media/music-by/mtime_y/2004/acid_jazz/akiko_kohara/2003-ak_note
    /usr/media/music-by/mtime_ym/200410/acid_jazz/akiko_kohara/2003-ak_note
    /usr/media/music-by/mtime_ymd/20041023/acid_jazz/akiko_kohara/2003-ak_note
    /usr/media/music-by/mtime_y_m/2004/10_october/acid_jazz/akiko_kohara/2003-ak_note
    /usr/media/music-by/mtime_y_m_d/2004/10_october/23_saturday/acid_jazz/akiko_kohara/2003-ak_note

This gives a very useful way to find albums in the music archive. For example,
"by genre and date" links can be used to see what progressive rock albums have
been released in 1979; "by artist" links can be used to see all of the artist's
albums even if he/she have released albums in several genres; "by mtime" links
can be used to see the albums in the order they were added to the music archive
(this is very useful to see the latest additions to the music archive).

Config files: 'music.conf', 'hardlink.conf'

2.3. C programs

Unfortunately, some features not provided by shell had to
be implemented in C.

2.3.1. cr2lf, lf0

Text filters. Change CRs to LFs, LFs to NULs respectively.

2.3.2. gettimeofday, isatty, nanosleep

The gettimeofday(2), isatty(3), nanosleep(2) shell counterparts (respectively).

2.3.3. readdef

Similar to bash's 'read -e', with a few extras.
Also uses the readline library.
Allows to supply a default value and edit history.

2.3.4. tracksplit

Splits a raw audio file to several wave audio files as
specified by offsets in the offsets file.

2.3.5. cdbgetall

Prints all records with a given key found in a constant database.
The 'constant database' here pertains to Dan Bernstein's cdb software.
The cdbgetall itself is derived from cdbget from cdb-0.75.

2.4. Configuration files

The most of the ova programs can be more or less configured.
The config files are shell scripts, and are interpreted at start-up.
Options are documented in config files themselves.

2.5. General notes

The proposed initial use of ova is the following:
  [if you don't have/want to use vorbisgain, remove 'v' from compound actions]

  - move all your music to $music_incoming (default: /usr/media/music/incoming)
  - split all single-file albums to separate tracks using ova action '0'

  - re/encode all tracks to Ogg Vorbis using ova action '2'
  - identify Ogg Vorbis tracks using ova action '3'
  - tag&relocate Ogg Vorbis tracks using ova compound action '4v'
  [optionally, do it using ova compound action '234v']

  - run ova-lint on your $music_home to check tags / find incomplete albums

At this point, your $music_incoming should become empty, and your
music archive should become very nicely classified.
Actually, you may also need to do the following:

  - for each non-MP3 single-file album (FLAC, APE, MPC etc.), decode it to cd-quality
      raw file (no wave header) using your favorite software
      example for ape:
        mac 1.ape 1.wav && mplayer 1.wav -ao pcm:nowaveheader:file=1.raw </dev/null && rm 1.ape 1.wav
    then use tracksplit to split the raw file to individual wave files
      example:
        ova-cue2offsets 1.cue >offsets && tracksplit 1.raw offsets . && rm 1.raw
      example:
        ova-cddb_read blues 330dce14 -o >offsets && tracksplit 1.raw offsets . && rm 1.raw
  - decode non-MP3 albums (FLAC, APE, MPC etc.), to wave files
      example for mpc:
        for i in *.mpc; do mppdec "$i" $(echo "$i" | sed 's/\.mpc$/.wav/') && rm "$i"; done
  - run ova action '234'/'234v' again

The proposed further use of ova is the following:

  - rip new CDs using ova compound action '1234v'.
  - split new single-file albums to separate tracks using ova action '0'
  - process new MP3/wave albums using ova compound action '234v'

  - run ova-lint on your $music_home to check tags / find incomplete albums

Advanced features:

  - use ova action 'h' (or ova-hardlink script) to make hard links for your
      albums, this will make them accessible via different paths, making
      it easy to find albums in the music archive

You are also encouraged to make submissions to the CDDB (freedb) by
using ova action 's' (the number of entries pending for submission
can be viewed using ova action 'i').

2.6. Reporting bugs / Sending patches / Requesting enhancements

If you think you've found a bug, written a patch or want a new
feature badly, feel free to mail me (see the head of this file).
Be sure to set the subject to 'ova: bugreport: blah blah blah',
'ova: patch: blah blah blah' or 'ova: rfe: blah blah blah'
for bugreport, patch or RFE respectively.
When writing bugreports, please include as much information as
possible. It's also imperative that you set 'enable_debug' in
'global.conf' to yes, reproduce your bug and check non-empty files
in 'log/error' subdirectory in ova's datadir (by default it's
/usr/local/share/ova) - their contents might alone let you fix
the bug. In any case, send these error log files along with the list
of commands/actions you've done that produced these error log files.


NEWS.md
1.3.0 on Mon, 26 Sep 2005 15:16:29 +0400
  NEW: bin/hardlink.sh
  bin/hardlink.sh: Makes hard links for an album / all albums.
  ova: Support for user configuration files ($HOME/.ova/*.conf), as
    requested by my SO.
  bin/sed.sh: Added "-a|--add=STRING" option (allows to add STRING
    to file's comments.
  bin/sed.sh: Added support for processing multiple files.
  bin/sed.sh: When output is a tty, mark changed files with '*'.
  bin/sed.sh: When output is a tty, apply sed script(s) to filename,
    if it's changed, move the old file to the new filename.
  bin/main.sh: Log splitted tracks in @datadir@/log/ova.tracksplit.
  bin/*.sh: Delete empty error log files on clean exit.
  share/global: Removed sed + utf-8 locale workaround (it caused
    more troubles than good).
  bin/*.sh, share/global: Use stat instead of "ls|..." to determine
    files' size.
  bin/main.sh: Fixed some minor bugs (with 'misc' detection).
  README: Updated docs.

1.2.0 on Mon, 29 Aug 2005 12:27:36 +0400
  ova: Support for CDDB protocol level 6 (database entries
    are now in UTF-8)
  ova: use MPlayer-1.0pre7 option "-ao pcm:file=FILENAME"
  ova: use "-n" option together with head and tail
  install: don't overwrite existing configuration files: instead,
    install new files (if changed) to $name.conf.new.
  INSTALL: vorbis-tools >= 1.1.1 is required; patching is no
    longer necessary.
  INSTALL: audiofile >= 0.2.5 is required; patching is no
    longer necessary.
  INSTALL: mp3tools >= 1.5 is required; patching is no
    longer necessary.
  INSTALL: Requirement for djb's cdb.
  NEW: src/cdbgetall.tar.gz/README
  bin/gettag_ogg.sh, bin/main.sh: Use existing REPLAYGAIN
    tags from Ogg Vorbis files if present.
  bin/main.sh: Consequently, the vorbisgain action applies
    to files in 'incoming' as well (it was pruned before
    because the tags would have been deleted on tag&relocate).
  bin/main.sh: Sort dirlisting in identify, reverse-sort
    dirlisting in tag&relocate (reverse because deep-to-shallow
    directory order is needed), sort dirlisting in replaygain.
  NEW: bin/sed.sh
  bin/sed.sh: Processes a given .ogg file's comments with sed.
  NEW: bin/lint.sh
  bin/lint.sh: Checks for files with missing tags and for
    missing tracks in the middle of albums.
  NEW: bin/cue2offsets.sh
  bin/cue2offsets.sh: Converts a cue sheet to offsets file
    readable by tracksplit.
  src/tracksplit.c: Seek to the first offset in the input
    file instead of the beginning of this file.
  src/tracksplit.c: Added '--seek=FRAMES' option. If it's
    given, it's argument is used instead of the first offset
    for seeking in the input file.
  bin/main.sh: Use tracksplit with '--seek=0' option.
  README: Added description for ova-lint. Compound actions
    ending with 'v' documented. Added descriptions for
    ova-cue2offsets, ova-sed.
  etc/main.conf: Automatically disable use_eject if the
    'eject' command is not found in PATH.

1.1.6 on Wed, 24 Dec 2003 22:59:05 +0300
  bin/main.sh: identify: Don't output short GPL notice when
    editing info in the EDITOR (it's irrelevant there).
  etc/global.conf: enable_debug is set to no by default for
    I believe that ova is stable enough for this.
  bin/main.sh: encode: Handle QUIT as abort when current
    jobs finish. Also, do nothing if queue is already empty.
  bin/main.sh: encode: Display total decoded/encoded
    playlength progress.
  bin/main.sh: rip: Correct detecting of already ripped,
    encoded, identified and tagged&relocated tracks.
  bin/main.sh: Added compound actions ending with 'v'.
  bin/main.sh: Change CD discid list in tag&relocate. This
    feature is very important, but was forgotten until now.
  bin/main.sh: identify: EDITOR editing: don't output
    copyright information and single quote example.
  bin/main.sh: tag_relocate: Change leading dots to 'dot-'.
  bin/main.sh: Added '4v' compound action.

1.1.5 on Sat, 20 Dec 2003 15:59:13 +0300
  bin/main.sh: encode: Delete '.decoding', '.encoding' and
    '.error' files on user abort.
  bin/main.sh: encode: Finish current jobs and quit upon
    receiving the QUIT signal.
  bin/conv.sh, bin/main.sh: Validate only decoding (assume
    encoding can't fail).
  bin/main.sh: identify: leave only track_artist, track_title
    and (if present) genre for 'misc' directories.
  etc/music.conf: track_path_pattern: Pattern for misc.
    Slightly changed other patterns (allow for empty artist).
  bin/main.sh: encode: Fixed multiple enqueueing.
  bin/main.sh, bin/cddb_number.sh: Better tracknumber matching.

1.1.4 on Tue, 16 Dec 2003 01:22:26 +0300
  bin/main.sh: encode: Continue on errors; report them in
    the error log file and barf at encoding completion.
  bin/main.sh: encode: Fixed total progress overflowing when
    encoding a large number of tracks (will still overflow
    if the number is sufficiently large, but it's unlikely
    that your HDD is big enough for this to happen ;).
  bin/main.sh: encode: Skip MP3s for which playlength
    couldn't be determined (probably broken headers).
  bin/conv.sh: Do sync after decoding and encoding.
  share/global: Fixed stripspaces.
  bin/main.sh: encode: Don't use '.status' and '.progress'
    files, use '.decoding' and '.encoding' instead.

1.1.3 on Sun, 14 Dec 2003 03:35:56 +0300
  bin/main.sh: split_tracks: Create output directory right
    after reading the CDDB category and discid.
  src/tracksplit.c: Fixed bug with improper rounding of the
    file size fractional part when displaying information.
  NEWS: Edited comment regarding bin/gettag_ogg.sh in 1.1.2.
  README: Cosmetic changes, synced main features from webpage.

1.1.2 on Sat, 13 Dec 2003 15:37:50 +0300
  bin/cddb_ibuild.sh: Workaround for one kind of CDDB entries
    that clearly violate CDDB entry format specification, but
    nevertheless are present in the latest freedb.
  bin/gettag_ogg.sh: Get average bitrate instead of nominal.
    I'm still not absolutely sure which one should be taken.

src/tracksplit.c
/*
 * copyright 2003-2005 Gene Pavlovsky <gene.pavlovsky@gmail.com>
 *
 * this is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * tracksplit: split a raw audio file to several wave audio files as
 *   specified by offsets in the offsets file.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>

#define FRAME_SIZE 2352  /* 1 frame of 44100 Hz, 2 channel, 16 bit audio */
#define FRAMES_PER_SECOND 75

#define P_MEMCPY_BLOCKSIZE 1048576

#define CL1 "\033[37m"
#define RST "\033[0m"

void p_memcpy(void *dest, void *src, size_t bytes)
{
  int done = 0, chunk;
  double progress;
  while ((chunk = bytes - done) > 0) {
    if (chunk > P_MEMCPY_BLOCKSIZE)
      chunk = P_MEMCPY_BLOCKSIZE;
    memcpy((char *)dest + done, (char *)src + done, chunk);
    done += chunk;
    progress = (double)done * 100 / bytes;
    printf("\033[7D"CL1"%3d.%.1d"RST"%% ", (int)floor(progress),
      (int)floor((progress - floor(progress)) * 10));
    fflush(stdout);
  }
}

void usage(char * name)
{
  fprintf(stderr, "Usage: %s [--seek=FRAMES] audio_file.raw offset_file output_dir\n", name);
  fprintf(stderr, "\nSplits the audio file according to the track frame offsets file,\n");
  fprintf(stderr, "writing the output wave files to the selected directory.\n");
  fprintf(stderr, "Track frame offsets file consists of a sequence of track frame\n");
  fprintf(stderr, "offsets for each track, separated by any amount of whitespace.\n");
  fprintf(stderr, "If the --seek option is given, it's argument is used instead of the\n");
  fprintf(stderr, "first offset for seeking to the initial position in the input file.\n");
  exit(2);
}

int main(int argc, char ** argv)
{
  FILE * offset_file;
  char * offset_line, * outfile_name;
  int * offsets;
  int first_offset, seek_offset = -1;
  int offset_size, i, trackcount, str_size, audiofile_size, audio_pos, audio_size, flag;
  int audiofile, outfile;
  int argidx = 0;
  double playlen;
  void * in_map, * out_map;
  char * plural;
  char wave_header[45] = "\122\111\106\106\000\000\000\000\127\101\126\105\146"
    "\155\164\040\020\000\000\000\001\000\002\000\104\254\000\000\020\261\002"
    "\000\004\000\020\000\144\141\164\141\000\000\000\000";

  if (argc <= 3)
    usage(argv[0]);

  if (strstr(argv[1], "--seek=")) {
    ++argidx;
    sscanf(argv[1] + 7, "%ld", &seek_offset);
  }

  if (argc <= 3 + argidx)
    usage(argv[0]);

  if (!(offset_file = fopen(argv[2 + argidx], "r"))) {
    perror("Error opening offset file");
    exit(1);
  }

  fseek(offset_file, 0, SEEK_END);
  offset_size = ftell(offset_file);
  fseek(offset_file, 0, SEEK_SET);
  offset_line = (char *)malloc(offset_size + 1);

  if (!fread((void *)offset_line, offset_size, 1, offset_file)) {
    if (ferror(offset_file)) {
      perror("Error reading offset file");
      exit(1);
    } else {
      fprintf(stderr, "No track frame offsets in offset file\n");
      exit(1);
    }
  }
  trackcount = 0;
  i = 0;
  while ((i < offset_size) && (isspace(offset_line[i]) || (offset_line[i] == '\n')))
    ++i;
  for (; i < offset_size; ++i)
    if (isdigit(offset_line[i])) {
      ++trackcount;
      while ((i < offset_size) && (isdigit(offset_line[i])))
        ++i;
      while ((i < offset_size) && (isspace(offset_line[i]) || (offset_line[i] == '\n')))
        ++i;
    }
  free(offset_line);
  if (trackcount <= 0) {
    fprintf(stderr, "No track frame offsets in offset file\n");
    exit(1);
  }

  offsets = (int *)malloc(sizeof(int) * trackcount);
  fseek(offset_file, 0, SEEK_SET);
  for (i = 0; i < trackcount; ++i)
    flag = fscanf(offset_file, "%ld", offsets + i);
    if (!flag || (flag == EOF)) {
      fprintf(stderr, "Error parsing offset file\n");
      exit(1);
    }
  fclose(offset_file);
  if (seek_offset >= 0)
    first_offset = seek_offset;
  else
    first_offset = offsets[0];

  for (i = 0; i < trackcount - 1; ++i)
    offsets[i] = offsets[i + 1] - offsets[i];
  offsets[trackcount - 1] = 0;

  if ((audiofile = open(argv[1 + argidx], O_RDONLY)) == -1) {
    perror("Error opening raw file");
    exit(1);
  }
  audiofile_size = lseek(audiofile, 0, SEEK_END);
  lseek(audiofile, 0, SEEK_SET);
  if ((in_map = mmap(0, audiofile_size, PROT_READ, MAP_SHARED, audiofile, 0)) == MAP_FAILED) {
    perror("Error mmaping raw file");
    exit(1);
  }
  close(audiofile);

  plural = trackcount > 1 ? "tracks" : "track";
  playlen = (double)audiofile_size / (FRAME_SIZE * FRAMES_PER_SECOND);
  printf("raw file: "CL1"%2d"RST":"CL1"%02d"RST" in "CL1"%3d.%d"RST" MiB, contains "CL1"%d"RST" %s\n",
    (int)floor(playlen) / 60, (int)floor(playlen) % 60, audiofile_size / 1048576,
    (int)floor((double)(audiofile_size % 1048576)/104857.6), trackcount, plural);

  str_size = strlen(argv[3 + argidx]) + strlen("/track??.wav") + 1;
  outfile_name = (char *)malloc(str_size);
  audio_pos = FRAME_SIZE * first_offset;
  for (i = 0; i < trackcount; ++i) {
    audio_size = audiofile_size - audio_pos;
    if (((i < trackcount - 1) && (audio_size < FRAME_SIZE * offsets[i])) || (audio_size <= 0)) {
      fprintf(stderr, "Raw file ends prematurely\n");
      exit(1);
    }
    if ((i < trackcount - 1) && (FRAME_SIZE * offsets[i] < audio_size))
      audio_size = FRAME_SIZE * offsets[i];
    snprintf(outfile_name, str_size, "%s/track%02d.wav", argv[3 + argidx], i + 1);
    playlen = (double)audio_size / (FRAME_SIZE * FRAMES_PER_SECOND);
    printf("track "CL1"%02d"RST": "CL1"%2d"RST":"CL1"%02d"RST" in "CL1"%3d.%d"RST" MiB, writing: "CL1"  0.0%% "RST, i + 1,
      (int)floor(playlen) / 60, (int)floor(playlen) % 60, (audio_size + 44) / 1048576,
      (int)floor((double)((audio_size + 44) % 1048576)/104857.6));
    fflush(stdout);
    if ((outfile = open(outfile_name, O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) == -1) {
      perror("Error opening output file");
      exit(1);
    }
    *((unsigned long *)(wave_header + 4)) = audio_size + 36; /* 44 - 8 */
    *((unsigned long *)(wave_header + 40)) = audio_size;
    write(outfile, wave_header, 44);
    lseek(outfile, audio_size + 43, SEEK_SET);
    write(outfile, "", 1);
    lseek(outfile, 0, SEEK_SET);
    if ((out_map = mmap(0, audio_size + 44, PROT_WRITE, MAP_SHARED, outfile, 0)) == MAP_FAILED) {
      perror("Error mmaping output file");
      exit(1);
    }
    close(outfile);
    p_memcpy((char *)out_map + 44, (char *)in_map + audio_pos, audio_size);
    munmap(out_map, audio_size + 44);
    audio_pos += audio_size;
    printf("\033[7D\033[K done\n");
  }

  munmap(in_map, audiofile_size);
  free(outfile_name);
  free(offsets);
}
